back/prisma/scripts/denormalize-form-sirets.ts

```diff
@@ -0,0 +1,76 @@
+import { Status } from "@prisma/client";
+import prisma from "../../src/prisma";
+import { registerUpdater, Updater } from "./helper/helper";
+
+@registerUpdater(
+  "Denormalize sirets",
+  `Populate forms sirets after schema denormalization`,
+  true
+)
+export class DenormalizeFormSiretsUpdater implements Updater {
+  async run() {
+    console.info("Starting denormalization script...");
+
+    try {
+      const listOfStatus = Object.values(Status);
+      console.info("1/3 Populating recipientsSirets...");
+
+      for (const status of listOfStatus) {
+        console.info(`Processing status = ${status} batch`);
+        await prisma.$executeRawUnsafe(`
```

Comment:
Une raison particulière de faire ça dans un script plutôt que dans un fichier de migration ?

Reply:
J'ai peur que le fichier de migration ne fase qu'un seul commit à la fin. Tout es joué dans la même transaction. Pour des raisons de perfs, mieux vaut séparer en plusieurs commits je pense. Sinon on risque de locker toute la table, et sur la durée des 3 requêtes

Reply:
Mais pour moi tous les fichiers sont dans la même transaction, non ? Puisque si une fail tout est rollback

Reply:
D'après moi à chaque fichier sa transaction, si une migration passe et que la suivante ne passe pas, la première migration n'est pas rollback.

![Capture d’écran 2022-11-24 à 10 15 10](https://user-images.githubusercontent.com/2269165/203744286-9488082e-6396-4f7e-85d8-7439f643f0d0.png)

Reply:
Et faut qu'on discute du déploiement.
Ce que j'imagine:
- créer les colonnes à la main sans index
- jouer le script de migration
- puis ensuite merger le code, il ignorera les colonnes déjà créées, ajoutera les indexes, et commencera à taper sur des données pré remplies.

Je viens de modifier les `add column` en `add column if not exists` pour permettre ça

Reply:
Si tu fais tout avec des fichiers de migration tu maitrise l'ordre du coup.

Reply:
Ah je n'avais pas vu ta réponse. Alors on peut faire ça:
1. Création de colonnes
2. Script 1
3. Script 2
4. Script 3
5. Création d'indexes

Reply:
La seule diff, c'est que pour éviter de faire toute la table d'un coup, je loopais sur les statuts. Pas certain qu'on puisse faire ce split en sql dans la même transaction

Reply:
Le truc c'est que même si tu boucles sur les statuts vu que c'est en séquentiel ça va quand même lock la table complètement il me semble. Donc autant faire tout d'un coup.

Reply:
Tu es sûr ? J'aurais dit que ça ne lock que les rows updated moi, non ?
Pour moi c'est un read sequentiel (et encore y'a un index sur le statut), et un lock appliqué uniquement aux rows effectivement updatés

Reply:
À voir ouais, c'est juste que ce système de script de migration ce n'est pas très robuste. Il faut penser à le faire tourner sur tous les environnements, à le désactiver ensuite, etc. Si tu veux batch ça doit aussi être possible en dupliquant un peu de code dans plusieurs fichiers de migrations. Plutôt que de batch par statut, on pourrait faire un premier batch des bordereaux updaté récemment, puis un second batch avec tout le reste.

Reply:
J'ai vu qu'on peut désactiver les transactions avec un commentaire en début de fichier (dans le cadre de cette migration de données, on s'en fiche de rollback).
J'ai peut être un truc pas mal là, dis moi ce que tu en penses

Reply:
Si ça fonctionne comme ça c'est top ouais

Reply:
Ca a fonctionné en local sur un dump de sandbox, et Laurent a pu le jouer sur un dump de prod sur sa machine (14 min de temps total).
Donc ca semble ok, je garde ça